apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using Google.Protobuf;
+using System;
+
+namespace Google.Cloud.Bigtable.V2
+{
+    /// <summary>
+    /// Keeps track of Rows returned from a readRows RPC for information relevant to resuming the RPC
+    /// after temporary problems.
+    /// </summary>
+    internal class BigtableReadRowsRequestManager
+    {
+        /// <summary>
+        /// Compares two <see cref="ByteString"/>s
+        /// </summary>
+        /// <returns>integer that indicates their relative position in the lexicographical order.</returns>
+        private int ByteStringCompare(ByteString key1, ByteString key2)
+        {
+            int size1 = key1.Length;
+            int size2 = key2.Length;
+            int size = Math.Min(size1, size2);
+
+            for (int i = 0; i < size; i++)
+            {
+                int comparison = key1[i].CompareTo(key2[i]);
+                if (comparison != 0)
+                {
+                    return comparison;
+                }
+            }
+            return size1.CompareTo(size2);
+        }
+
+        // Memeber variables form the constructor.
+        private readonly ReadRowsRequest _originalRequest;
+
+        // The number of rows read so far.
+        private long _rowCount;
+
+        /// <summary>
+        /// Constructor for ResumingStreamingResultScanner.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containg all of the parameters of the API call.
+        /// </param>
+        public BigtableReadRowsRequestManager(ReadRowsRequest originalRequest)
+        {
+            _originalRequest = originalRequest;
+        }
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        public ByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        public long RowCount
+        {
+            get => _rowCount;
+            set => _rowCount += value;
```

Comment:
Using the property setter as an increment instead of a normal set is will not be expected by callers and might cause some maintenance issues. I suggest making this a readonly property: `internal long RowCount { get; private set; }` and then having a separate `IncrementRowCount(long value)` method.